### PR TITLE
feat(poetry): Add private package index support to Poetry

### LINKED ISSUE #6488: Python private package index support for poetry
**What would you like Renovate to be able to do?**

I would like Renovate to be able to update `poetry.lock` files that contain dependencies from a private package index.

Similar to #4487, you can add a `hostRule`, such as

```
{
      'baseUrl': 'https://my-private-pypi.com/',
      'username': 'username',
      'password': 'password'
}
```

to the `config.js` file. This allows Renovate to find outdated dependencies, but not **update them**. The result is that the `pyproject.toml` file gets updated, whereas the `poetry.lock` file does not.

From the log files, you can extract this error message:

```
Creating virtualenv venv in /home/ubuntu/.cache/pypoetry/virtualenvs
Updating dependencies
Resolving dependencies...

[SolverProblemError]
Because X depends on Y (^0.Z.0) which doesn't match any versions, version solving failed.
```

I believe this has to do with Renovate using the `renovate/python` Docker image when trying to [update artifacts](https://github.com/renovatebot/renovate/blob/master/lib/manager/poetry/artifacts.ts#L79). For poetry to be able to update dependencies from a private package index, it first needs to be configured, as described [here](https://python-poetry.org/docs/repositories/#using-a-private-repository).

**Describe the solution you'd like**

There should be a way to configure poetry from the `config.js` so that you can specify private package indices.

**Describe alternatives you've considered**

You should be able to use the global pre-configured poetry on your system, instead of the `renovate/python` Docker image.

### PR DESCRIPTION
**Pull request by jophish**
This PR adds support for private Poetry sources by comparing sources' `url` field in `pyproject.toml` against host rules. If a match is found for a source `url`, Renovate will look for the `username` and `password` fields in the matching host rule, construct appropriate environment variables expected by Poetry (based off of the source's `name` field), and pass them to the execution of `poetry update --lock`.

Closes #6488.

### CODE REVIEW

File: lib/manager/poetry/artifacts.ts [hunk 1]
```diff
@@ -76,8 +109,11 @@ export async function updateArtifacts({
     const tagConstraint = getPythonConstraint(existingLockFileContent, config);
     const poetryRequirement = config.compatibility?.poetry || 'poetry';
     const poetryInstall = 'pip install ' + quote(poetryRequirement);
+    const sourceCredentialVars = await getSourceCredentialVars(packageFileName);
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.ts:112]: Pass `newPackageFileContent` instead, it is the contents of the `pyproject.toml`

File: lib/manager/poetry/artifacts.ts [hunk 2]
```diff
@@ -37,6 +39,37 @@ function getPythonConstraint(
...
+): Promise<{ [s: string]: string }> {
+  const pyprojectFileName = getSiblingFileName(
+    packageFileName,
+    'pyproject.toml'
+  );
+  const pyprojectContent = await readLocalFile(pyprojectFileName, 'utf8');
+  if (!pyprojectContent) {
+    logger.debug(`No pyproject.toml found`);
+    return {};
+  }
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.ts:53]:
```suggestion
  pyprojectContent: string
): Promise<{ [s: string]: string }> {
```

File: lib/manager/poetry/extract.ts [hunk 3]
```diff
@@ -87,6 +87,32 @@ function extractRegistries(pyprojectfile: PoetryFile): string[] {
   return Array.from(registryUrls);
 }
 
+export function getPoetrySources(
```
Comment (rarkins) [inline lib/manager/poetry/extract.ts:90]: Relocate this function to the artifacts file

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/poetry/artifacts.ts (hunk 1 above)
Comment (Jophish) [reply lib/manager/poetry/artifacts.ts:112]: :facepalm: Thanks -- didn't realize that!

**Comment by Jophish** [comment]
I wasn't sure if `getPoetrySources` should mimic the pattern [here](https://github.com/renovatebot/renovate/blob/master/lib/manager/poetry/extract.ts#L92) of accepting a `fileName` argument solely for logging purposes. I've copied that behavior for now, which means passing a `packageFileName` argument to `getSourceCredentialVars`.

File: lib/manager/poetry/artifacts.ts [hunk 4]
```diff
@@ -37,6 +39,50 @@ function getPythonConstraint(
...
+    logger.debug({ err }, 'Error parsing pyproject.toml file');
+    return [];
+  }
+  if (!pyprojectFile.tool?.poetry) {
+    logger.debug(`{$fileName} contains no poetry section`);
+    return [];
+  }
+
+  const sources = pyprojectFile.tool?.poetry?.source;
+  const sourceSet = new Set<PoetrySource>();
```
Comment (rarkins) [inline lib/manager/poetry/artifacts.ts:56]: Does using a Set here provide any advantage versus regular array? As it contains objects, I don't think we'll get any de-duplication?

**Review by rarkins — COMMENTED** [review]
LGTM, just one comment re: Set vs array

File: lib/manager/poetry/artifacts.ts (hunk 4 above)
Comment (Jophish) [reply lib/manager/poetry/artifacts.ts]: Oops -- you're right, we won't get dedpulication. Changed it to an array.

File: lib/manager/poetry/artifacts.ts [hunk 5]
```diff
@@ -37,6 +39,50 @@ function getPythonConstraint(
...
+  }
+  return sourceArray;
+}
+
+function getSourceCredentialVars(
+  pyprojectContent: string,
+  packageFileName: string
+): { [s: string]: string } {
+  const poetrySources = getPoetrySources(pyprojectContent, packageFileName);
+  const envVars = {};
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:70]:
```suggestion
  const envVars: { [s: string]: string } = {};
```

File: lib/manager/poetry/artifacts.ts [hunk 6]
```diff
@@ -76,8 +122,14 @@ export async function updateArtifacts({
     const tagConstraint = getPythonConstraint(existingLockFileContent, config);
     const poetryRequirement = config.compatibility?.poetry || 'poetry';
     const poetryInstall = 'pip install ' + quote(poetryRequirement);
+    const sourceCredentialVars = getSourceCredentialVars(
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:125]:
```suggestion
    const extraEnv = getSourceCredentialVars(
```

File: lib/manager/poetry/artifacts.ts [hunk 7]
```diff
@@ -76,8 +122,14 @@ export async function updateArtifacts({
...
+      newPackageFileContent,
+      packageFileName
+    );
+
     const execOptions: ExecOptions = {
       cwdFile: packageFileName,
+      extraEnv: sourceCredentialVars,
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:132]:
```suggestion
      extraEnv,
```

**Review by viceice — CHANGES_REQUESTED** [review]
just some small improvements

File: lib/manager/poetry/artifacts.ts (hunk 5 above)
Comment (Jophish) [reply lib/manager/poetry/artifacts.ts:70]: Thanks for this suggestion; haven't used typescript before so not sure what kind of syntax is required 🙂

File: lib/manager/poetry/artifacts.ts [hunk 8]
```diff
@@ -37,6 +39,50 @@ function getPythonConstraint(
...
+      sourceArray.push({ name: source.name, url: source.url });
+    }
+  }
+  return sourceArray;
+}
+
+function getSourceCredentialVars(
+  pyprojectContent: string,
+  packageFileName: string
+): { [s: string]: string } {
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:68]:
```suggestion
): Record<string, string> {
```

File: lib/manager/poetry/artifacts.ts [hunk 9]
```diff
@@ -37,6 +39,50 @@ function getPythonConstraint(
...
+  const poetrySources = getPoetrySources(pyprojectContent, packageFileName);
+  const envVars: { [s: string]: string } = {};
```
Comment (viceice) [inline lib/manager/poetry/artifacts.ts:70]:
```suggestion
  const envVars: Record<string, string> = {};
```

**Review by viceice — COMMENTED** [review]
just more simplification, 🙈  Will merge them myself

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
Thanks for your PR, @Jophish !

**Pull request merged**